subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy

```diff
@@ -187,67 +183,247 @@ class JavaToolchainBuildOperationsIntegrationTest extends AbstractIntegrationSpe
+
+        buildFile << """
+            abstract class ToolchainTask extends DefaultTask {
+                @Nested
+                abstract Property<JavaLauncher> getLauncher1()
+                @OutputFile
+                abstract RegularFileProperty getOutputFile()
+
+                @TaskAction
+                void myAction() {
+                    def output = outputFile.get().asFile
+                    output << launcher1.get().executablePath
+                }
+            }
 
-    // TODO: javaCompile.options.forkOptions.javaHome = "blah" while using a toolchain
+            tasks.register("myToolchainTask1", ToolchainTask) {
+                outputFile = layout.buildDirectory.file("output.txt")
+                launcher1 = javaToolchains.launcherFor {
+                    languageVersion = JavaLanguageVersion.of(${jdkMetadata1.languageVersion.majorVersion})
+                }
+            }
 
-    // TODO: test with Kotlin plugin
+            tasks.register("myToolchainTask2", ToolchainTask) {
+                outputFile = layout.buildDirectory.file("output.txt")
+                launcher1 = javaToolchains.launcherFor {
+                    languageVersion = JavaLanguageVersion.of(${jdkMetadata2.languageVersion.majorVersion})
+                }
+            }
+        """
 
-    def "emit toolchain usages when task fails for 'compileJava' task"() {
-        def task = "compileJava"
-        def taskPath = ":$task"
+        def installationPaths = [jdkMetadata1, jdkMetadata2].collect { it.javaHome.toAbsolutePath().toString() }.join(",")
+
+        when:
+        runWithInstallationPaths(installationPaths, ":myToolchainTask1", ":myToolchainTask2")
+        def events1 = eventsFor(":myToolchainTask1")
+        def events2 = eventsFor(":myToolchainTask2")
+        then:
+        executedAndNotSkipped(":myToolchainTask1", ":myToolchainTask2")
+        assertToolchainUsages(events1, jdkMetadata1, "JavaLauncher")
+        assertToolchainUsages(events2, jdkMetadata2, "JavaLauncher")
+
+        // TODO: should it work with up-to-date tasks?
+//        when:
+//        runWithInstallationPaths(installationPaths, ":myToolchainTask1", ":myToolchainTask2")
+//        events1 = eventsFor(":myToolchainTask1")
+//        events2 = eventsFor(":myToolchainTask2")
+//        then:
+//        skipped(":myToolchainTask1", ":myToolchainTask2")
+//        assertToolchainUsages(events1, jdkMetadata1, "JavaLauncher")
+//        assertToolchainUsages(events2, jdkMetadata2, "JavaLauncher")
+    }
+
+    def "emits toolchain usages for compilation that configures java home via fork options"() {
+        JvmInstallationMetadata jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentJdk)
+
+        buildFile << """
+            compileJava {
+                options.forkOptions.javaHome = file("${jdkMetadata.javaHome}")
```

Comment:
I think it would be interesting to run with having one toolchain for the project (`runWithInstallation`) and another one for the `javaHome`. Would it then emit two different toolchain events?